### PR TITLE
Add an analyzer that warns about non-literal sequence numbers

### LINKED ISSUE #14579: Analyzer to detect nonconstant sequence number in RenderTreeBuilder calls
It's always wrong to have a nonconstant value for the `sequence` parameter to `RenderTreeBuilder` methods, so we should have an analyzer that fails the build if you do.

(In case anyone's wondering why, [here's why](https://gist.github.com/SteveSandersonMS/ec232992c2446ab9a0059dd0fbc5d0c3))

### PR DESCRIPTION
**Pull request by MackinnonBuck**
**Add an analyzer that warns about non-literal sequence numbers**
The analyzer warns when using anything other than integer literals for sequence numbers in `RenderTreeBuilder` methods.

**Potential improvements to consider**
There are definitely situations where it's appropriate for developers to use non-literals as the sequence number (e.g. `RenderTreeBuilder` extension methods that accept a `sequence` and pass it to a `RenderTreeBuilder` method). In such cases, should we expect developers to suppress the warning, or should we attempt something smarter? For example, if the argument passed as the sequence number is itself a parameter of the containing method, should we perform the same analysis on the method containing the `RenderTreeBuilder` method invocation? Should we go even further and ensure that the parameter is not reassigned anywhere in the containing method?

Fixes #14579

### CODE REVIEW

**Review by pranavkm — COMMENTED** [review]
The changes look great. I'd prefer if @SteveSandersonMS to sign off on this to see if there was more / less we should be doing here.

File: src/Framework/Analyzer/src/DiagnosticDescriptors.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        "IActionResult instances should not be returned from a {0} Delegate parameter. Consider returning an equivalent result from Microsoft.AspNetCore.Http.Results.",
+        "Usage",
+        DiagnosticSeverity.Warning,
+        isEnabledByDefault: true,
+        helpLinkUri: "https://aka.ms/aspnet/analyzers");
+
+    internal static readonly DiagnosticDescriptor DoNotUseNonLiteralSequenceNumbers = new(
+        "ASP0005",
+        "Do not use non-literal sequence numbers",
+        "'{0}' should not be used as a sequence number",
```
Comment (SteveSandersonMS) [inline src/Framework/Analyzer/src/DiagnosticDescriptors.cs:32]:
```suggestion
        "'{0}' should not be used as a sequence number. Instead, use an integer literal representing source code order.",
```

File: src/Framework/Analyzer/test/Components/DisallowNonLiteralSequenceNumbersTest.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+    [Fact]
+    public async Task RenderTreeBuilderInvocationWithNonConstantArgument_ProducesDiagnostics()
+    {
+        // Arrange
+        var source = TestSource.Read(@"
+using Microsoft.AspNetCore.Components.Rendering;
+var renderTreeBuilder = new RenderTreeBuilder();
+var i = 0;
+renderTreeBuilder.OpenRegion(/*MM*/i);
```
Comment (SteveSandersonMS) [inline src/Framework/Analyzer/test/Components/DisallowNonLiteralSequenceNumbersTest.cs:38]: Does the `/*MM*/` play a functional role here? Is this to test that the analyzer is looking for the first argument, and not the first bit of syntax within the parens?

Comment (SteveSandersonMS) [reply src/Framework/Analyzer/test/Components/DisallowNonLiteralSequenceNumbersTest.cs:38]: If we're keeping this, would it be OK to change it to something more descriptive like `/* Comment that the analyzer will skip */`?

File: src/Framework/Analyzer/src/DiagnosticDescriptors.cs (hunk 1 above)
Comment (SteveSandersonMS) [reply src/Framework/Analyzer/src/DiagnosticDescriptors.cs:32]: ... unless you end up finding some way to link to docs about this instead.

**Review by SteveSandersonMS — APPROVED** [review]
Looks brilliant - thanks!

File: src/Framework/Analyzer/test/Components/DisallowNonLiteralSequenceNumbersTest.cs (hunk 2 above)
Comment (pranavkm) [reply src/Framework/Analyzer/test/Components/DisallowNonLiteralSequenceNumbersTest.cs:38]: Our test infrastructure uses this to figure out where the analyzer produces the diagnostics at. It specifically looks for this token: https://github.com/dotnet/aspnetcore/blob/main/src/Analyzers/Microsoft.AspNetCore.Analyzer.Testing/src/TestSource.cs#L12

File: src/Framework/Analyzer/src/RenderTreeBuilder/DiagnosticDescriptors.cs [hunk 3]
```diff
@@ -0,0 +1,19 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.CodeAnalysis;
+
+namespace Microsoft.AspNetCore.Analyzers.RenderTreeBuilder;
```
[the remark this answers is not in the document]
Comment (pranavkm) [reply src/Framework/Analyzer/src/RenderTreeBuilder/DiagnosticDescriptors.cs]: Just this one should be fine.

**Pull request merged**